src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerSession.cs

```diff
@@ -1,201 +1,336 @@
-        /// Is the session currently being polled
-        /// </summary>
-        public bool IsPolling
-        {
-            get
-            {
-                return this.isPolling;
-            }
-            set
-            {
-                lock (this.pollingLock)
-                {
-                    this.isPolling  = value;
-                }
-            }
-        }
+        /// Underlying XEvent session wrapper
+        /// </summary>
+        public IXEventSession XEventSession => xEventSession;
+
+        /// <summary>
+        /// Try to set the session into polling mode if criteria is meet
+        /// </summary>
+        /// <returns>True if session set to polling mode, False otherwise</returns>
+        public bool TryEnterPolling()
+        {
+            lock (this.pollingLock)
+            {
+                if (pollImmediately || (!this.isPolling && DateTime.Now.Subtract(this.lastPollTime) >= PollingDelay))
+                {
+                    this.isPolling = true;
+                    this.lastPollTime = DateTime.Now;
+                    this.pollImmediately = false;
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Is the session currently being polled
+        /// </summary>
+        public bool IsPolling
+        {
+            get
+            {
+                return this.isPolling;
+            }
+            set
+            {
+                lock (this.pollingLock)
+                {
+                    this.isPolling  = value;
+                }
+            }
+        }
+
+        /// <summary>
+        /// The delay between session polls
+        /// </summary>
+        public TimeSpan PollingDelay { get; } = DefaultPollingDelay;
+
+        /// <summary>
+        /// Could events have been lost in the last poll
+        /// </summary>
+        public bool EventsLost
+        {
+            get
+            {
+                return this.eventsLost;
+            }
+        }
+
+        /// <summary>
+        /// Determine if an event was caused by the XEvent polling queries
+        /// </summary>
+        private bool IsProfilerEvent(ProfilerEvent currentEvent)
+        {
+            if (string.IsNullOrWhiteSpace(currentEvent.Name) ||  currentEvent.Values == null)
+            {
+                return false;
+            }
+
+            if ((currentEvent.Name.Equals("sql_batch_completed")
+                || currentEvent.Name.Equals("sql_batch_starting"))
+                && currentEvent.Values.TryGetValue("batch_text", out string value))
+            {
+                return value.Contains("SELECT target_data FROM sys.dm_xe_session_targets")
+                    || currentEvent.Values["batch_text"].Contains("SELECT target_data FROM sys.dm_xe_database_session_targets");
+            }
+
+            return false;
+        }

```

Comment:
it'd be nice to remove this event filter eventually. The SSMS model is to use whatever events the session has defined; SSMS XEvent profiler will create the initial session, but the user is free to modify it as they like. We can't really presume that this limited set of events is adequate for a very broad range of investigations.